### DISCUSSION TITLE
Keep receiving syntax errors in series of conditionals, beginner Python

### DISCUSSION THREAD

**Discussion by retrohp** [community/community#21471] [Other Feature Feedback, Questions, & Ideas]
<p>I’m attempting to solve a palindromic puzzle with Python that requires me to have the following conditionals</p>
<ol>
<li>The number is 6 digits long</li>
<li>The last 4 digits are palindromic</li>
<li>One mile later last 5 digits are palindromic</li>
<li>One mile later middle 4 digits are palindromic</li>
<li>One mile later all 6 digits are palindromic If all these conditionals are fulfilled the puzzle is solved The full problem is found <a href="https://www.cartalk.com/radio/puzzler/palindromic-odometer" rel="noopener nofollow ugc">here</a> if you need more context</li>
</ol>
<p>I want to solve this problem using only a  <code>while</code>  loop, slicing, and  <code>if</code>  statements, I haven’t learned about  <code>def</code>  statements yet.</p>
<p>Right now I have:</p>
<pre><code class="lang-auto">winner = False
while x &lt; 1000000 and winner: 
    if x_str = str(x) :
       x_str[2:6] == x[5:1:-1] 
    if x1_str = str(x+1) :
       x1_str[1:6] == x1[-1:0:-1]  
    if x2_str = str(x+2) :
       x2_str[1:5] == x2[-2:-6:-1] 
    if x3_str = str(x+3) : 
       x3_str[0:6] == x3[-1:-7:-1] 
    if winner(True):
       print ("Great! The answer is", x) 
</code></pre>
<p>But I get a syntax error on my first conditional. If I change the assignment operator to == instead, then I get undefined error for all my variables. I’m not sure what to do here? I’ve only been coding for two weeks so it might be a very simple error.</p>
<p>Thanks so much</p>

**Comment by airtower-luna** [comment, marked as answer]
<aside class="quote no-group" data-username="retrohp" data-post="1" data-topic="134404">
<div class="title">
<div class="quote-controls"></div>
 retrohp:</div>
<blockquote>
<pre><code class="lang-python">    if x_str = str(x) :
</code></pre>
</blockquote>
</aside>
<p>A conditional must contain a condition (a value that is <code>True</code> or <code>False</code>, or can be converted to one of those). An assignment  is no condition, so you get a syntax error.</p>
<p>You always need to assign a variable before you can use it. A variable that hasn’t been assigned doesn’t exist, that’s why you get an error saying it’s undefined.</p>
<p>I’m not sure <em>why</em> you’re trying to do an assignment in the <code>if</code> condition here, it looks the thing you put in the body of the conditional is what you actually want to be the condition. If you want to assign <code>x_str</code> and then compare its slices, do this:</p>
<pre><code class="lang-python">    x_str = str(x)
    if x_str[2:6] == x_str[5:1:-1]:
        # Do what should happen if the condition is True
</code></pre>
<p>I’m seeing a few more issues here:</p>
<aside class="quote no-group" data-username="retrohp" data-post="1" data-topic="134404">
<div class="title">
<div class="quote-controls"></div>
 retrohp:</div>
<blockquote>
<pre><code class="lang-python">winner = False
while x &lt; 1000000 and winner: 
</code></pre>
</blockquote>
</aside>
<ol>
<li>The loop body will never be executed, because <code>winner</code> is <code>False</code> right from the start.</li>
<li>
<code>x</code> is undefined. I assume you want to check all possible <code>x</code> within a certain range, so you’ll either have to initialize it with a small value and increment it in the loop, or (much simpler) use a <a href="https://docs.python.org/3/tutorial/controlflow.html#for-statements"><code>for</code> loop</a> with a <a href="https://docs.python.org/3/tutorial/controlflow.html#the-range-function"><code>range()</code> expression</a>.</li>
</ol>
<p>Finally, if you use a <code>for</code> loop (and thus don’t have to increment anything in your code), you can just skip to the next number using a <a href="https://docs.python.org/3/tutorial/controlflow.html#break-and-continue-statements-and-else-clauses-on-loops"><code>continue</code> statement</a> as soon as one of the checks fails. <code>break</code> (explained in the same section) might be useful to stop when you find a winning number.</p>
